### PR TITLE
Fix SSR static build public file copying. fixes #3016

### LINKED ISSUE #3016: 🐛 BUG: build copying public folder for build is not recursive
### What version of `astro` are you using?

latest

### What package manager are you using?

npm

### What operating system are you using?

linux (wsl)

### Describe the Bug

With https://stackblitz.com/edit/github-4qjebp?file=astro.config.mjs&on=stackblitz

dev server is fine but "build" fails when copying the public folder with:

"ENOENT: no such file or directory, copyFile '/home/projects/github-4qjebp/public/assets/logo.svg' -> '/home/projects/github-4qjebp/dist/client/assets/logo.svg'"

Reason is [CopyFiles](https://github.com/withastro/astro/blob/1335797903a57716e9a02b0ffd8ca636b3883c62/packages/astro/src/core/build/static-build.ts#L251-L266) does not create the required subfolder (eg like "mkdir -p")

Sorry not minimal but this is where I hit the problem. I checked on local too. Build makes little in stackblitz but currently that's the only way to create a SSR endpoint. I hope this will change so can just "dev" without "build" step. Dito Netlify functions so can run locally

### Link to Minimal Reproducible Example

 https://stackblitz.com/edit/github-4qjebp?file=astro.config.mjs&on=stackblitz

### Participation

- [x] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by SteveALee**
## Changes

- When SSR is enabled and it's a static build the copying of public files is broken with sub folders
- - fix is to create subfolders as required

## Testing

Test added to static -build.test.js

## Docs

none

### CODE REVIEW

File: packages/astro/src/core/build/static-build.ts [hunk 1]
```diff
@@ -253,15 +253,14 @@ async function copyFiles(fromFolder: URL, toFolder: URL) {
...
-	// Make the directory
-	await fs.promises.mkdir(toFolder, { recursive: true });
-
 	await Promise.all(
 		files.map(async (filename) => {
 			const from = new URL(filename, fromFolder);
 			const to = new URL(filename, toFolder);
-			return fs.promises.copyFile(from, to);
-		})
+			const toSubPath = new URL(npath.dirname(filename), toFolder)
```
Comment (SteveALee) [inline packages/astro/src/core/build/static-build.ts:260]: So I'm not sure why URL() was used. Semantically confusion but perhaps used as quick platform independent way to to join paths? Anyway I followed the style.

Result is a tad inefficient as for every file will travel the entire path from file system root creating required folders.

**Comment by SteveALee** [comment]
I meant to say that an alternative is to use something like `ncp` module but didn't fancy an extra dependency sucking in loads of code address edge case and options we don't have. #tradeoffs

File: packages/astro/src/core/build/static-build.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:260]: Correct, we use `file://` `URL`s internally as a platform independent way to represent any path, then convert them back to paths as needed using `fileURLToPath(urlValue)`.

Comment (natemoo-re) [reply packages/astro/src/core/build/static-build.ts:260]: My gut reaction is that `new URL('../', filename)` would be better than `npath.dirname(filename)` here, but I'm not positive without taking this branch for a spin locally.

**Review by natemoo-re — COMMENTED** [review]
Looks good! Not sure about using `npath.dirname` with a URL, but I'll let @matthewp weigh in.

**Comment by SteveALee** [comment]
[natemoo-re](https://github.com/natemoo-re) yeah i think that would be fine and better fits the rest of the code. I just used what I'd used before in path manipulations

File: packages/astro/src/core/build/static-build.ts (hunk 1 above)
Comment (SteveALee) [reply packages/astro/src/core/build/static-build.ts]: hah! We were all fooled it's actually the to file's folder, ie './' :)

**Review by natemoo-re — APPROVED** [review]
Nice! Thank you for the PR 👍🏻

**Comment by SteveALee** [comment]
@matthewp So The function "async function ssrMoveAssets(opts: StaticBuildOptions)" would appear to have the same issue. Except - I suspect only the root folder needs to be moved, not all the sub files/folders.

**Comment by FredKSchott** [comment]
LGTM! 

In the future, try to avoid source files/folders with spaces in the name, it can be a nightmare to type out or copy-paste into the terminal. I'll update for you post-merging the PR.

**Comment by SteveALee** [comment]
@FredKSchott oh gosh, I introduced a filename bug? Sorry. I'm at a loss to how that happened - I usually use kebab case and thought I just copied the existing test file. My only excuse is a new monitor has defaulted to smaller text and I thought I could cope - lol

But thanks for accepting - I'm well chuffed to have contributed. Huzzah!

**Comment by FredKSchott** [comment]
No worries! I think you did copy an existing test, which I hadn't realized when I commented. No worries at all :)

**Pull request merged**